kratos/solving_strategies/schemes/residual_based_bdf_custom_scheme.h

```diff
@@ -0,0 +1,588 @@
+            }
+        }
+    }
+
+    /** Copy Constructor.
+     */
+    ResidualBasedBDFCustomScheme(ResidualBasedBDFCustomScheme& rOther)
+        :BDFBaseType(rOther)
+        ,mDoubleVariable(rOther.mDoubleVariable)
+        ,mFirtsDoubleDerivatives(rOther.mFirtsDoubleDerivatives)
+        ,mSecondDoubleDerivatives(rOther.mSecondDoubleDerivatives)
+        ,mArrayVariable(rOther.mArrayVariable)
+        ,mFirtsArrayDerivatives(rOther.mFirtsArrayDerivatives)
+        ,mSecondArrayDerivatives(rOther.mSecondArrayDerivatives)
+    {
+    }
+    
+    /**
+     * Clone
+     */
+    BaseTypePointer Clone() override
+    {
+        return BaseTypePointer( new ResidualBasedBDFCustomScheme(*this) );
+    }
+
+    /** Destructor.
+     */
+    ~ResidualBasedBDFCustomScheme
+    () override {}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Performing the prediction of the solution
+     * @details It predicts the solution for the current step x = xold + vold * Dt
+     * @param rModelPart The model of the problem to solve
+     * @param rDofSet set of all primary variables
+     * @param A LHS matrix
+     * @param Dx Incremental update of primary variables
+     * @param b RHS Vector
+     */
+
+    void Predict(
+        ModelPart& rModelPart,
+        DofsArrayType& rDofSet,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b
+        ) override
+    {
+        KRATOS_TRY;
+
+        ProcessInfo& current_process_info = rModelPart.GetProcessInfo();
+        const double delta_time = current_process_info[DELTA_TIME];
+
+        // Updating time derivatives (nodally for efficiency)
+        const int num_nodes = static_cast<int>( rModelPart.Nodes().size() );
+
+        #pragma omp parallel for
+        for(int i = 0;  i< num_nodes; ++i) {
+            auto it_node = rModelPart.Nodes().begin() + i;
+
+            //ATTENTION::: the prediction is performed only on free nodes
+            std::size_t counter = 0;
+            for ( const auto& i_var : mDoubleVariable) {
+                // Derivatives
+                const Variable<double> dvar = mFirtsDoubleDerivatives[counter];
+                const Variable<double> d2var = mSecondDoubleDerivatives[counter];
+                
+                // Values
+                const double dot2un1 = it_node->FastGetSolutionStepValue(d2var, 1);
+                const double dotun1 = it_node->FastGetSolutionStepValue(dvar, 1);
+                const double un1 = it_node->FastGetSolutionStepValue(i_var, 1);
+                const double dot2un0 = it_node->FastGetSolutionStepValue(d2var);
+                double& dotun0 = it_node->FastGetSolutionStepValue(dvar);
+                double& un0 = it_node->FastGetSolutionStepValue(i_var);
+                
+                if (it_node->HasDofFor(d2var)) {
```

Comment:
These ifs are repeated several times. Is it possible to put them in a private template method?

Reply:
Okidoki

Reply:
also observe that all the nodes have the same dofs.

so if the first node has the dof, all the others will (this allows doing one single check instead of once per node)

Reply:
Done